### ISSUE TITLE
[no-inferrable-types] useful type annotation is reported as error

### ISSUE THREAD

**Issue by pgsandstrom** [typescript-eslint/typescript-eslint#3752]
- [ X] I have tried restarting my IDE and the issue persists.
- [ X] I have updated to the latest version of the packages.
- [ X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```TS
type doSomethingType<T extends (arg: any) => void> = (callback: T) => void

const doSomething: doSomethingType<(n: any) => void> = () => {
  //
}

doSomething((myVar:number = 1) => {
  // no-inferrable-types reports an error here. But if I remove type annotation, then myVar is of type 'any'
})
```

**Expected Result**

Since the type annotation actually changes the type, then it should not be reported as an error.

**Actual Result**

It is reported as an error.

**Additional Info**

The real world situation where I ran into this was when using reacts useCallback hook.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.2` |
| `@typescript-eslint/parser`        | `4.29.2` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `15.3.0` |

**Comment by bradzacher** [comment]
Duplicate of #3145
